### PR TITLE
DG-1370 Write Mesh parent attributes in server

### PR DESCRIPTION
**Pull request by nikhilbonte21**
## Change description
We have the following de-normalized for mesh assets:

parentDomainQualifiedName → stores qualifiedName of the immediate parent domain

superDomainQualifiedName → stores qualifiedName of the root domain

These attributes are managed by client, this can lead to inconsistencies like we saw for [JPMC](https://atlanhq.atlassian.net/browse/DG-1338).

Writing these values from BE itself will help us avoid inconsistencies between actual relationships & these de-normalized attributes.

## Type of change
- [x] Bug fix (fixes an issue)
- [ ] New feature (adds functionality)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### Code review 

- [ ] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [ ] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

### CODE REVIEW

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DataProductPreProcessor.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        RequestContext.get().endMetricRecord(metricRecorder);
+    }
+
+    private AtlasEntityHeader getNewParentDomain(AtlasEntity entity, EntityMutationContext context) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("ProductPreProcessor.getNewParentDomain");
+
+        AtlasObjectId objectId = (AtlasObjectId) entity.getRelationshipAttribute(PRODUCT_DOMAIN_REL_TYPE);
+
```
Comment (PRATHAM2002-DS) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DataProductPreProcessor.java:112]: The object returned by entity.getRelationshipAttribute(PRODUCT_DOMAIN_REL_TYPE) is instanceOf Map and not AtlasObjectId in my opinion. Can you please check it once.

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DataProductPreProcessor.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            entity.removeAttribute(SUPER_DOMAIN_QN);
+            if (entity.getRelationshipAttributes() != null) {
+                entity.getRelationshipAttributes().remove(PRODUCT_DOMAIN_REL_TYPE);
+            }
+        }
+
+        RequestContext.get().endMetricRecord(metricRecorder);
+    }
+
+    private AtlasEntityHeader getNewParentDomain(AtlasEntity entity, EntityMutationContext context) throws AtlasBaseException {
```
Comment (PRATHAM2002-DS) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DataProductPreProcessor.java:108]: Can you please rename this function to setParent as we are using this name in other features.

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DomainPreProcessor.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            entity.removeAttribute(SUPER_DOMAIN_QN);
+            if (entity.getRelationshipAttributes() != null) {
+                entity.getRelationshipAttributes().remove(DOMAIN_DOMAIN_PARENT_REL_TYPE);
+            }
+        }
+
+        RequestContext.get().endMetricRecord(metricRecorder);
+    }
+
+    private AtlasEntityHeader getNewParentDomain(AtlasEntity entity, EntityMutationContext context) throws AtlasBaseException {
```
Comment (PRATHAM2002-DS) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DomainPreProcessor.java:131]: Can you please rename this function to setParent as we are using this name in other features.

**Review by PRATHAM2002-DS — APPROVED** [review]

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DataProductPreProcessor.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+    private void processUpdateProduct(AtlasEntity entity, AtlasVertex vertex, EntityMutationContext context) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("processUpdateProduct");
+
+        AtlasEntity storedProduct = entityRetriever.toAtlasEntity(vertex);
+        AtlasObjectId currentParent = (AtlasObjectId) storedProduct.getRelationshipAttribute(PRODUCT_DOMAIN_REL_TYPE);
+
+        AtlasEntityHeader newParent = getNewParentDomain(entity, context);
+
+        if (currentParent == null && newParent != null) {
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DataProductPreProcessor.java:84]: What happens if currentParent is not null?

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DataProductPreProcessor.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+            String superDomainQualifiedName = (String) newParent.getAttribute(SUPER_DOMAIN_QN);
+            if (StringUtils.isEmpty(superDomainQualifiedName)) {
+                String[] splitted = parentDomainQualifiedName.split("/");
+                superDomainQualifiedName = String.format("%s/%s/%s", splitted[0], splitted[1], splitted[2]);
+            }
+
+            entity.setAttribute(SUPER_DOMAIN_QN, superDomainQualifiedName);
+
+        } else {
+            //never update these attributes with API request
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DataProductPreProcessor.java:99]: We should return an error here.

**Pull request closed without merging**